### PR TITLE
crimson: Set device class during spawn of a crimson osd

### PR DESCRIPTION
**Pull request by mohit84**
Implement a wrapper for different backend storage to set device_class during spawn of a process

Fixes: https://tracker.ceph.com/issues/66627

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by mohit84** [comment]
jenkins test make check

File: src/crimson/osd/osd.cc [hunk 1]
```diff
@@ -608,6 +610,41 @@ seastar::future<> OSD::_send_boot()
...
+    return seastar::now();
+  }
+
+  auto get_device_class = [this]() -> seastar::future<std::string> {
+    return store.get_default_device_class().then([FNAME, this](std::string device_class) {
+      return seastar::make_ready_future<std::string>(device_class);
+    }).handle_exception([FNAME](std::exception_ptr e) {
+      ERROR("Failed to get device class: ", e);
+      return seastar::make_ready_future<std::string>("");
+    });
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:626]: Are we expecting an thrown exception here?
For expected errors in Crimson we use errorator, see dev/crimson/error-handling.

File: src/crimson/osd/osd.cc [hunk 2]
```diff
@@ -608,6 +610,41 @@ seastar::future<> OSD::_send_boot()
...
+  };
+
+  return get_device_class().then([FNAME, this](auto device_class) {
+    INFO("device_class is {} whoami: {}", device_class, whoami);
+    string cmd = string("{\"prefix\": \"osd crush set-device-class\", ") +
+                 string("\"class\": \"") + device_class + string("\", ") +
+                 string("\"ids\": [\"") + stringify(whoami) + string("\"]}");
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:633]: Let's switch to fmt::format here

File: src/crimson/osd/osd.cc [hunk 3]
```diff
@@ -608,6 +610,41 @@ seastar::future<> OSD::_send_boot()
...
+  auto get_device_class = [this]() -> seastar::future<std::string> {
+    return store.get_default_device_class().then([FNAME, this](std::string device_class) {
+      return seastar::make_ready_future<std::string>(device_class);
+    }).handle_exception([FNAME](std::exception_ptr e) {
+      ERROR("Failed to get device class: ", e);
+      return seastar::make_ready_future<std::string>("");
+    });
+  };
+
+  return get_device_class().then([FNAME, this](auto device_class) {
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:629]: We should check that `device_class` is not an empty string before using it

File: src/crimson/osd/osd.cc [hunk 4]
```diff
@@ -608,6 +610,41 @@ seastar::future<> OSD::_send_boot()
...
+    INFO("device_class is {} whoami: {}", device_class, whoami);
+    string cmd = string("{\"prefix\": \"osd crush set-device-class\", ") +
+                 string("\"class\": \"") + device_class + string("\", ") +
+                 string("\"ids\": [\"") + stringify(whoami) + string("\"]}");
+
+    return monc->run_command(std::move(cmd), {}).then([FNAME](auto&& command_result) {
+      [[maybe_unused]] auto [code, message, out] = std::move(command_result);
+      if (code) {
+        WARN("fail to set device_class : {} ({})", message, code);
+        throw std::runtime_error("fail to add to crush");
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:639]: "fail to set device_class"

**Review by Matan-B — COMMENTED** [review]
looks good so far, left a few comments.
I would also suggest trying to use coroutines in `OSD::_add_device_class()` as it can make the implementation easier to read.

**Comment by Matan-B** [comment]
makecheck: 
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/osd.cc:621:58: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
    return store.get_default_device_class().then([FNAME, this](std::string device_class) {
                                                  ~~~~~~~^~~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/crimson/osd/osd.cc:623:26: error: variable 'FNAME' cannot be implicitly captured in a lambda with no capture-default specified
    }).handle_exception([FNAME](std::exception_ptr e) {
```

File: src/crimson/osd/osd.cc [hunk 5]
```diff
@@ -608,6 +610,41 @@ seastar::future<> OSD::_send_boot()
   return monc->send_message(std::move(m));
 }
 
+seastar::future<> OSD::_add_device_class()
+{
+  LOG_PREFIX(OSD::_add_device_class);
+  if (!local_conf().get_val<bool>("osd_crush_update_on_start")) {
```
[the remark this answers is not in the document]
Comment (mohit84) [reply src/crimson/osd/osd.cc:616]: done

File: src/crimson/osd/osd.cc (hunk 2 above)
Comment (mohit84) [reply src/crimson/osd/osd.cc:633]: done

File: src/crimson/osd/osd.cc (hunk 3 above)
Comment (mohit84) [reply src/crimson/osd/osd.cc:629]: done

File: src/crimson/osd/osd.cc (hunk 4 above)
Comment (mohit84) [reply src/crimson/osd/osd.cc:639]: done

File: src/crimson/osd/osd.cc (hunk 1 above)
Comment (mohit84) [reply src/crimson/osd/osd.cc:626]: I have changed the code now, please let me know if it is ok now.

**Comment by mohit84** [comment]
jenkins test make check

**Comment by mohit84** [comment]
jenkins test make check

File: src/crimson/osd/osd.cc [hunk 6]
```diff
@@ -608,6 +610,52 @@ seastar::future<> OSD::_send_boot()
...
+
+  auto get_device_class = [FNAME, this]() -> seastar::future<std::string> {
+    try {
+      std::string device_class = co_await store.get_default_device_class();
+      co_return device_class;
+    } catch (const std::exception& e) {
+      ERROR("Failed to get device class: ", e.what());
+      co_return "";
+    }
+  };
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:628]: There's no need for try/catch we discarding exceptions as we handle error with Crimson's errorator concept.
See `TEST_F(coroutine_test_t, test_ertr_coroutine_error)`

Comment (Matan-B) [reply src/crimson/osd/osd.cc:628]: We can skip the `get_device_class` wrapper: 
```
std::string device_class = co_await store.get_default_device_class();
```

File: src/crimson/osd/osd.cc [hunk 7]
```diff
@@ -608,6 +610,52 @@ seastar::future<> OSD::_send_boot()
...
+      if (code) {
+        WARN("fail to set device_class : {} ({})", message, code);
+        throw std::runtime_error("fail to set device_class");
+      } else {
+        INFO("newly added to crush: {}", message);
+      }
+  } catch (const std::exception& e) {
+    ERROR("Failed to add device class: ", e.what());
+    throw;
+  }
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:654]: ditto

File: src/crimson/osd/osd.cc [hunk 8]
```diff
@@ -608,6 +610,52 @@ seastar::future<> OSD::_send_boot()
...
+    }
+
+    INFO("device_class is {} whoami: {}", device_class, whoami);
+
+    std::string cmd = fmt::format(
+      R"({{"prefix": "osd crush set-device-class", "class": "{}", "ids": ["{}"]}})",
+      device_class, stringify(whoami)
+    );
+
+    auto [code, message, out] = co_await monc->run_command(std::move(cmd), {});
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:644]: Great usage!

File: src/crimson/osd/osd.cc [hunk 9]
```diff
@@ -608,6 +610,52 @@ seastar::future<> OSD::_send_boot()
...
+      if (code) {
+        WARN("fail to set device_class : {} ({})", message, code);
+        throw std::runtime_error("fail to set device_class");
```
Comment (Matan-B) [inline src/crimson/osd/osd.cc:647]: nit:
```
// to be caught by crimson/osd/main.cc 
throw std::runtime_error("fail to set device_class");
```

File: src/crimson/osd/osd.cc (hunk 6 above)
Comment (mohit84) [reply src/crimson/osd/osd.cc]: done

File: src/crimson/osd/osd.cc (hunk 7 above)
Comment (mohit84) [reply src/crimson/osd/osd.cc]: I have changed the code

File: src/crimson/osd/osd.cc (hunk 9 above)
Comment (mohit84) [reply src/crimson/osd/osd.cc]: done

File: src/crimson/osd/osd.cc [hunk 10]
```diff
@@ -608,6 +610,38 @@ seastar::future<> OSD::_send_boot()
...
+    co_return;
+  }
+
+  std::string device_class = co_await store.get_default_device_class();
+  if (device_class.empty()) {
+    WARN("Device class is empty; skipping crush update.");
+    co_return;
+  }
+
+  INFO("device_class is {} whoami: {}", device_class, whoami);
```
[the remark this answers is not in the document]
Comment (mohit84) [reply src/crimson/osd/osd.cc]: done

**Comment by mohit84** [comment]
jenkins test make check

**Comment by mohit84** [comment]
@Matan-B 
https://pulpito.ceph.com/moagrawa-2024-11-20_12:34:54-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/

It seems 2 jobs are failed, look;s like the failed job is similar to below tracker
https://tracker.ceph.com/issues/68883
https://tracker.ceph.com/issues/20960

**Comment by mohit84** [comment]
> @Matan-B https://pulpito.ceph.com/moagrawa-2024-11-20_12:34:54-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/
> 
> It seems 2 jobs are failed, look;s like the failed job is similar to below tracker https://tracker.ceph.com/issues/68883 https://tracker.ceph.com/issues/20960

Failed
https://pulpito.ceph.com/moagrawa-2024-11-20_12:34:54-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/8001232
https://pulpito.ceph.com/moagrawa-2024-11-20_12:34:54-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/8001268

Dead
https://pulpito.ceph.com/moagrawa-2024-11-20_12:34:54-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/8001240
https://pulpito.ceph.com/moagrawa-2024-11-20_12:34:54-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/8001243
https://pulpito.ceph.com/moagrawa-2024-11-20_12:34:54-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/8001274
https://pulpito.ceph.com/moagrawa-2024-11-20_12:34:54-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/8001279

**Comment by mohit84** [comment]
@Matan-B 
https://pulpito.ceph.com/moagrawa-2024-11-21_08:55:14-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/

Failures
8003283 https://tracker.ceph.com/issues/67972
8003305 https://tracker.ceph.com/issues/66852

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
> @Matan-B https://pulpito.ceph.com/moagrawa-2024-11-21_08:55:14-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/
> 
> Failures 8003283 https://tracker.ceph.com/issues/67972 8003305 https://tracker.ceph.com/issues/66852

Please also include "dead" jobs.
8003264 from perf suite failed as well, unrelated since a similar failure was seen in recent main run:
https://pulpito.ceph.com/teuthology-2024-11-23_20:56:02-crimson-rados-main-distro-crimson-smithi/ 

Merging, thank you!

**Comment by mohit84** [comment]
> @Matan-B https://pulpito.ceph.com/moagrawa-2024-11-21_08:55:14-crimson-rados-wip-mohit-crimson-device_class-distro-crimson-smithi/
> 
> Failures 8003283 https://tracker.ceph.com/issues/67972 8003305 https://tracker.ceph.com/issues/66852

 Dead job links, it seems both jobs are dead due to some issue at infra level.
  https://pulpito.ceph.com/teuthology-2024-11-23_20:56:02-crimson-rados-main-distro-crimson-smithi/8005981/
  https://pulpito.ceph.com/teuthology-2024-11-23_20:56:02-crimson-rados-main-distro-crimson-smithi/8005998/

**Comment by Matan-B** [comment]
jenkins test make check

**Comment by Matan-B** [comment]
> Dead job links, it seems both jobs are dead due to some issue at infra level. https://pulpito.ceph.com/teuthology-2024-11-23_20:56:02-crimson-rados-main-distro-crimson-smithi/8005981/ https://pulpito.ceph.com/teuthology-2024-11-23_20:56:02-crimson-rados-main-distro-crimson-smithi/8005998/

There seems to be something off with Crimson and cbt. As mentioned earlier these failures also appear in main so I'll merge once CI passes.

**Comment by Matan-B** [comment]
jenkins test make check

**Pull request merged**